# What is the input?

Text dataset that has alot of noise and false information and the existence of such data on the internet caused many social problems as it has a role on many different domains, the data needs alot of cleaning and preprocessing.

# What is the output?

 prediction if a specific post is fake or not, by looking at its title. the fake news' label is (0) and the real news will be labeled by (1)

# What data mining function is required?

binary classification with different models using text preprocessing techniques, pipelines,Random Search using a validation set

# What could be the challenges?

- In order to extract information from big data, we must ensure that our method is efficient and scalable, and we must have adequate knowledge and expertise to use them if we need to update our algorithms.

- Complex data that can be found in datasets.

- Creating a viable solution to our issue.

# What is the impact?

Our model predictions is going to predict wether a certain post is fake or not and this will greatly reduce the social problems caused by those fake news as many people believe them and they act according to them while in reality, they are not true at all

# What is an ideal solution?

In my perspective, an ideal solution will be measured in terms of metrics and results.

# What is the experimental protocol used and how was it carried out? What preprocessing steps are used?

when it comes to the preprocessing steps I made a function performing the following on the training set and testing set :

 remove any html tags, Keep only ASCII + European Chars and whitespace, remove single letter chars,convert all whitespaces  to single wspace.
 
if not for embedding :

all lowercase

remove stopwords, punctuation and stemm

stemming


The experimental protocol I used in my code is validation set

# What is the difference between Character n-gram and Word n-gram? Which one tends to suffer more from the OOV issue?

When comparing two linguistic versions, character N-grams or word N-grams. The alignment systems are traditionally based on word
N-grams splitting. The observation of the morphological variety of languages, even inside a single linguistic family, quickly shows
that the word granularity is inadequate to provide a widely multilingual system, i.e. a language independent system able to handle
flexional languages as well as positional languages. Instead, when starting from a multilingual collection to focus on pairs of texts,
we defend that character N-grams alignment is more efficient than word N-grams alignment

the word n-gram tends to suffer more from the OOV issue

# What is the difference between stop word removal and stemming? Are these techniques language-dependent?

Stop word removal is the process of deleting terms that appear often in all of the papers in the corpus. Articles and pronouns are typically categorised as stop words.

Stemming is the practise of reducing word inflection to its basic forms, such as mapping a group of words to the same stem, even if the stem isn't a valid word in the language.

both stemming and stop word removal are language-dependant

 # Is tokenization techniques language dependent? Why?

Tokenization is the process of breaking down a huge body of information into smaller lines, words, or even inventing new terms for a language other than English.

You should always tokenize document and query words the same way, whether you're using a Boolean or a free text inquiry, and process inquiries with the same tokenizer. This ensures that a string of characters in a text will always match a string of characters typed in a query.
These tokenization difficulties are language-specific. As a result, it is necessary to understand the document's language. It is quite successful to identify languages using classifiers that employ short character subsequences as features; most languages have distinct signature patterns.

# What is the difference between count vectorizer and tf-idf vectorizer? Would it be feasible to use all possible n-grams? If not, how should you select them?

The TfidfTransformer transforms a count matrix to a normalized tf or tf-idf representation. So although both the CountVectorizer and TfidfTransformer (with use_idf=False) produce term frequencies, TfidfTransformer is normalizing the count.

'n' is simply a variable with positive integer values such as 1,2,3, and so on. The letter 'n' stands for many. In a similar vein, depending on the value of 'n,' n-grams are categorised into the following categories.

1:Unigram

2:Bigram

3:Trigram

n:n-gram

No it won't be be feasible to use all possible n-grams

we need many different types of n-grams,This is because different types of n-grams are suitable for different types of applications. You should try different n-grams on your data in order to confidently conclude which one works the best among all for your text analysis.

A 1-gram (or unigram) is a one-word sequence

A2gram (or bigram) is a two-word sequence of words
 
Unigram can provide better test results than Bigram and Trigram

N-gram range sets if features to be used to characterize texts will be:

Unigrams or words (n-gram size = 1)

Bigrams or terms compounded by two words (n-gram size = 2)

Trigrams or terms compounded by up to three words (n-gram size = 3)



# Code

In [1]:
import pandas as pd #I imported pandas cause Using it
#make it way easier when it comes to  processing, analysis and manipulation of data
import numpy as np # as it has array Manipulation functions
from nltk.stem.snowball import SnowballStemmer#stemmer converts words to its root form
from nltk.tokenize import word_tokenize # we will be able to extract the tokens from string of characters
from nltk.corpus import stopwords # imported it to be able to remove stop words
import nltk #The Natural Language Toolkit which are libraries and programs for NLP

In [2]:
traind=pd.read_csv('xy_train.csv',index_col='id')#I imported my training data set
# and Set column "id" as index column
traind

text  label
id                                                              
265723  A group of friends began to volunteer at a hom...      0
284269  British Prime Minister @Theresa_May on Nerve A...      0
207715  In 1961, Goodyear released a kit that allows P...      0
551106  Happy Birthday, Bob Barker! The Price Is Right...      0
8584    Obama to Nation: 聙"Innocent Cops and Unarmed Y...      0
...                                                   ...    ...
70046   Finish Sniper Simo H盲yh盲 during the invasion o...      0
189377  Nigerian Prince Scam took $110K from Kansas ma...      1
93486   Is It Safe To Smoke Marijuana During Pregnancy...      0
140950  Julius Caesar upon realizing that everyone in ...      0
34509   Jeff Bridges Releasing 鈥楽leeping Tapes,鈥?a New...      1

[60000 rows x 2 columns]

In [3]:
testd=pd.read_csv('x_test.csv',index_col='id')
testd  #I imported my testing data set
# Set column "id" as index column

text
id                                                      
0                                             stargazer 
1                                                   yeah
2      PD: Phoenix car thief gets instructions from Y...
3      As Trump Accuses Iran, He Has One Problem: His...
4                           "Believers" - Hezbollah 2011
...                                                  ...
59146                  Bicycle taxi drivers of New Delhi
59147  Trump blows up GOP's formula for winning House...
59148  Napoleon returns from his exile on the island ...
59149   Deep down he always wanted to be a ballet dancer
59150  Toddler miraculously survives 6-story fall lan...

[59151 rows x 1 columns]

In [4]:
traind.head(10)#view first 10 rows

text  label
id                                                              
265723  A group of friends began to volunteer at a hom...      0
284269  British Prime Minister @Theresa_May on Nerve A...      0
207715  In 1961, Goodyear released a kit that allows P...      0
551106  Happy Birthday, Bob Barker! The Price Is Right...      0
8584    Obama to Nation: 聙"Innocent Cops and Unarmed Y...      0
117912  In the 1920鈥檚, Hitler was forbidden to address...      0
213064  Nerd Wins Scrabble with word you've never hear...      0
398923  Why 95.8% of Female Newscasters Have the Same ...      1
314798  Donald Trump Says He'll Do This If More 'Inapp...      0
20243   5 crazy facts about Lamborghini's outrageous e...      0

In [5]:
traind.isnull().any() #checking if there are any nulls in training set

text     False
label    False
dtype: bool

In [6]:
testd.isnull().any() #checking if there are any nulls in testing set

text    False
dtype: bool

In [7]:
nltk.download('punkt') #downloading the tokenizer which divides a text into a list of sentences from nltk
nltk.download('stopwords')#Stop words are words that are going to be ignored by tokenizers.

stemmer = SnowballStemmer("english") #stemmer converts words to its root form, for english language
stop_words = set(stopwords.words("english"))#to be able to remove stop words for english language

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Debi\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Debi\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [8]:
def clean_text(text, for_embedding=False): #defining a function for cleaning 
    """ steps:
        - remove any html tags (< /br> often found)
        - Keep only ASCII + European Chars and whitespace, no digits
        - remove single letter chars
        - convert all whitespaces (tabs etc.) to single wspace
        if not for embedding :
        - all lowercase
        - remove stopwords, punctuation and stemm
    """
    RE_WSPACE = re.compile(r"\s+", re.IGNORECASE) #for converting whitespacs to single white space
    RE_TAGS = re.compile(r"<[^>]+>") #for removing html tags
    RE_ASCII = re.compile(r"[^A-Za-zÀ-ž ]", re.IGNORECASE) #removing ASCII chars and european chars
    RE_SINGLECHAR = re.compile(r"\b[A-Za-zÀ-ž]\b", re.IGNORECASE) # removing single letter chars
    if for_embedding: #   if not for embedding :
        # - all lowercase
        # - remove stopwords, punctuation and stemm
        # Keep punctuation
        RE_ASCII = re.compile(r"[^A-Za-zÀ-ž,.!? ]", re.IGNORECASE)
        RE_SINGLECHAR = re.compile(r"\b[A-Za-zÀ-ž,.!?]\b", re.IGNORECASE)

    text = re.sub(RE_TAGS, " ", text)
    text = re.sub(RE_ASCII, " ", text)
    text = re.sub(RE_SINGLECHAR, " ", text)
    text = re.sub(RE_WSPACE, " ", text)

    word_tokens = word_tokenize(text) ## to extract the tokens from string of characters
    words_tokens_lower = [word.lower() for word in word_tokens]

    if for_embedding:
        # no stemming, lowering and punctuation / stop words removal
        words_filtered = word_tokens
    else:
        words_filtered = [
            stemmer.stem(word) for word in words_tokens_lower if word not in stop_words
        ]

    text_clean = " ".join(words_filtered)
    return text_clean

In [9]:
# Clean text
import re #The power of regular expressions is that they can specify patterns, not just fixed characters.

traind["text_clean"] = traind.loc[traind["text"].str.len() > 4, "text"] #defining a text clean column in training set
traind["text_clean"] = traind["text"].map( #applying the text_clean function on the 'text' column and putting the results in the text_clean column 
    lambda x: clean_text(x, for_embedding=False) if isinstance(x, str) else x
)

In [10]:
traind #diplaying the training dataset

text  label  \
id                                                                 
265723  A group of friends began to volunteer at a hom...      0   
284269  British Prime Minister @Theresa_May on Nerve A...      0   
207715  In 1961, Goodyear released a kit that allows P...      0   
551106  Happy Birthday, Bob Barker! The Price Is Right...      0   
8584    Obama to Nation: 聙"Innocent Cops and Unarmed Y...      0   
...                                                   ...    ...   
70046   Finish Sniper Simo H盲yh盲 during the invasion o...      0   
189377  Nigerian Prince Scam took $110K from Kansas ma...      1   
93486   Is It Safe To Smoke Marijuana During Pregnancy...      0   
140950  Julius Caesar upon realizing that everyone in ...      0   
34509   Jeff Bridges Releasing 鈥楽leeping Tapes,鈥?a New...      1   

                                               text_clean  
id                                                         
265723  group friend began volunt homeless shelter nei...  
284269  british prime minist theresa may nerv attack f...  
207715  goodyear releas kit allow ps brought heel http...  
551106  happi birthday bob barker price right host lik...  
8584    obama nation innoc cop unarm young black men d...  
...                                                   ...  
70046      finish sniper simo yh invas finland ussr color  
189377  nigerian princ scam took kansa man year later ...  
93486        safe smoke marijuana pregnanc surpris answer  
140950  julius caesar upon realiz everyon room knife e...  
34509   jeff bridg releas leep tape new album design h...  

[60000 rows x 3 columns]

In [11]:
testd["text_clean"] = testd.loc[testd["text"].str.len() > 4, "text"] #defining a text clean column in testing set
testd["text_clean"] = testd["text"].map(  #applying the text_clean function on the 'text' column and putting the results in the text_clean column
    lambda x: clean_text(x, for_embedding=False) if isinstance(x, str) else x
)

In [12]:
testd # displaying the testing dataset

text  \
id                                                         
0                                             stargazer    
1                                                   yeah   
2      PD: Phoenix car thief gets instructions from Y...   
3      As Trump Accuses Iran, He Has One Problem: His...   
4                           "Believers" - Hezbollah 2011   
...                                                  ...   
59146                  Bicycle taxi drivers of New Delhi   
59147  Trump blows up GOP's formula for winning House...   
59148  Napoleon returns from his exile on the island ...   
59149   Deep down he always wanted to be a ballet dancer   
59150  Toddler miraculously survives 6-story fall lan...   

                                            text_clean  
id                                                      
0                                              stargaz  
1                                                 yeah  
2       pd phoenix car thief get instruct youtub video  
3                 trump accus iran one problem credibl  
4                                     believ hezbollah  
...                                                ...  
59146                     bicycl taxi driver new delhi  
59147             trump blow gop formula win hous race  
59148  napoleon return exil island elba march colouris  
59149                    deep alway want ballet dancer  
59150       toddler miracul surviv stori fall land car  

[59151 rows x 2 columns]

In [13]:
traind.loc[traind["label"] > 1] = np.NaN #deleting any label that is not 0 or 1

# Drop when any of x missing
traind = traind[(traind["text_clean"] != "") & (traind["text_clean"] != "null")]

traind = traind.dropna(
    axis="index", subset=["label", "text", "text_clean"]
).reset_index(drop=True)

In [14]:
train_data_clean = traind.copy() #making a copy of the training data set
train_data_clean.head(3) #viewwing the first 3 rows

text  label  \
0  A group of friends began to volunteer at a hom...    0.0   
1  British Prime Minister @Theresa_May on Nerve A...    0.0   
2  In 1961, Goodyear released a kit that allows P...    0.0   

                                          text_clean  
0  group friend began volunt homeless shelter nei...  
1  british prime minist theresa may nerv attack f...  
2  goodyear releas kit allow ps brought heel http...

In [15]:
test_data_clean = testd.copy() #making a copy of the testing data set
test_data_clean.head(3)#viewwing the first 3 rows

text  \
id                                                      
0                                          stargazer    
1                                                yeah   
2   PD: Phoenix car thief gets instructions from Y...   

                                        text_clean  
id                                                  
0                                          stargaz  
1                                             yeah  
2   pd phoenix car thief get instruct youtub video

In [16]:
from bokeh.models import NumeralTickFormatter
# viewing Word Frequency of most common words
word_freq = pd.Series(" ".join(train_data_clean["text_clean"]).split()).value_counts()
word_freq[0:200]

year     4129
one      3285
like     3128
new      2998
look     2847
         ... 
star      565
ad        564
ago       563
came      563
elect     563
Length: 200, dtype: int64

In [17]:
# list most 5 uncommon words
word_freq[-5:].reset_index(name="freq")

index  freq
0  hatsun     1
1   nfler     1
2  hicock     1
3  mccall     1
4    wahr     1

In [18]:
# Distribution of ratings
train_data_clean["label"].value_counts(normalize=True)

0.0    0.538221
1.0    0.461779
Name: label, dtype: float64

In [19]:
#splitting the data into x and y for training data and x for testing data
Xtr=train_data_clean['text_clean']
Ytr=train_data_clean['label']
test_data_clean=testd['text_clean']

In [20]:
from sklearn.pipeline import Pipeline #Pipeline of transforms with a final estimator.
from sklearn.feature_extraction.text import TfidfVectorizer #to Convert text to a matrix of TF-IDF features.
from sklearn.linear_model import LogisticRegression #importing the logistic regression model


pipe = Pipeline([("tfidf", TfidfVectorizer("word")), ("lr", LogisticRegression())]) #we will use the word vectorizer and the logistic regrression model


params = {
    
    "tfidf__ngram_range": [(1, 4),(2,5)],#n-gram range:tuple (min_n, max_n), The lower and upper boundary of the range of n-values for different n-grams to be extracted. All values of n such that min_n <= n <= max_n will be used.
    "tfidf__max_df": np.arange(0.6, 0.8),#When building the vocabulary ignore terms that have a document frequency strictly higher than the given threshold (corpus-specific stop words).
    "tfidf__min_df": np.arange(5,30 ),#When building the vocabulary ignore terms that have a document frequency strictly lower than the given threshold. This value is also called cut-off in the literature. 
    'lr__penalty' : ['l1','l2'], #Penalized logistic regression imposes a penalty to the logistic model for having too many variables. This results in shrinking the coefficients of the less contributive variables toward zero. 
    'lr__C' : np.logspace(-3,3,30), #C value controls the strength of the penalty
    'lr__solver': [ 'liblinear','newton-cg', 'lbfgs'] #setting solver options/ranges to liblinear,newton-cg,lbfgs
}

C:\Users\Debi\anaconda3\lib\site-packages\sklearn\utils\validation.py:70: FutureWarning: Pass input=word as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error
  warnings.warn(f"Pass {args_msg} as keyword args. From version "


In [21]:
# now using random search with valisation set
#Further split the original training set to a train and a validation set
from sklearn.model_selection import train_test_split
from sklearn.model_selection import PredefinedSplit
from sklearn.model_selection import RandomizedSearchCV

X_train, X_val, y_train, y_val = train_test_split(
    Xtr, Ytr, train_size = 0.7, stratify = Ytr, random_state = 45)#the random_state parameter is used for initializing the internal random number generator
#The stratify option instructs sklearn to divide the dataset into a test and training set with the ratio of class labels in the variable supplied.

# Create a list where train data indices are -1 and validation data indices are 0
# X_train2 (new training set), X_train
split_index = [-1 if x in X_train.index else 0 for x in Xtr.index]

# Use the list to create PredefinedSplit
pds = PredefinedSplit(test_fold = split_index)
#
random_search_lr = RandomizedSearchCV(
    pipe, params, cv=pds, verbose=1, n_jobs=2, 
    # number of random trials
    n_iter=10,
    scoring='roc_auc')#The degree of separability/distinction or intermingling/crossover between the forecasts of the two classes is shown by the ROC-AUC.

random_search_lr.fit(Xtr,Ytr)  #fitting my data into random_search_lr results

print('best score {}'.format(random_search_lr.best_score_)) #printing best scores
print('best hyperparameter {}'.format(random_search_lr.best_params_)) #printing best hyperparameters

Fitting 1 folds for each of 10 candidates, totalling 10 fits


C:\Users\Debi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:922: UserWarning: One or more of the test scores are non-finite: [0.68153394 0.81905744 0.61759491        nan        nan        nan
 0.63161273 0.65247247 0.70316885        nan]
  warnings.warn(


best score 0.8190574376210747
best hyperparameter {'tfidf__ngram_range': (1, 4), 'tfidf__min_df': 25, 'tfidf__max_df': 0.6, 'lr__solver': 'liblinear', 'lr__penalty': 'l2', 'lr__C': 0.0016102620275609393}


In [22]:
submission = pd.DataFrame() # creating dataframe for the submission

submission['id'] = test_data_clean.index #using the column index of x test values to fill submission['id'] column

submission['label'] = random_search_lr.predict_proba(test_data_clean)[:,1] #predicting the probabilities and filling submission['label'] column with their values

submission.to_csv('predwalkthrough.csv', index=False)#generating the submission file


# Trial one using logistic regression classifier with random search and validation set

# Expectaion

I expect to get the optimal hyperparamters that will give me the best performance and highest accuracy using logistic regression model with random search and  validation set

# Observation 

After running the code,the best hyperparameters that was defined were : 
    
'tfidf__ngram_range': (1, 4)
    
'tfidf__min_df': 25
    
'tfidf__max_df': 0.6
    
'lr__solver': 'liblinear'
    
'lr__penalty': 'l2'
    
'lr__C': 0.0016102620275609393

kaggle score 0.83008

best score 0.8190574376210747

# Plan

I will change my vectorizer to character level vectorizer and for my hyperparameters, I will change their ranges to 

'lr__penalty' : [‘l1’, ‘l2’, ‘elasticnet’, ‘none’]

 'lr__C' : np.logspace(-6,6,40)
 
 'lr__solver': [ 'liblinear','newton-cg', ‘sag’, ‘saga’]

In [23]:
params = {
    
    "tfidf__ngram_range": [(1, 4),(2,5)], #n-gram range:tuple (min_n, max_n), The lower and upper boundary of the range of n-values for different n-grams to be extracted. All values of n such that min_n <= n <= max_n will be used.
    "tfidf__max_df": np.arange(0.6, 0.8), #When building the vocabulary ignore terms that have a document frequency strictly higher than the given threshold (corpus-specific stop words).
    "tfidf__min_df": np.arange(5,30 ),#When building the vocabulary ignore terms that have a document frequency strictly lower than the given threshold. This value is also called cut-off in the literature. 
    'lr__penalty' : ['l1', 'l2', 'elasticnet', 'none'],#Penalized logistic regression imposes a penalty to the logistic model for having too many variables. This results in shrinking the coefficients of the less contributive variables toward zero. 
    'lr__C' : np.logspace(-6,6,40), #C value controls the strength of the penalty
    'lr__solver': [ 'liblinear','newton-cg', 'sag', 'saga']#setting solver options/ranges to liblinear,newton-cg,lbfgs
}

In [24]:
random_search_lr = RandomizedSearchCV(
    pipe, params, cv=pds, verbose=1, n_jobs=2, 
    # number of random trials
    n_iter=10,
    scoring='roc_auc')#The degree of separability/distinction or intermingling/crossover between the forecasts of the two classes is shown by the ROC-AUC.

random_search_lr.fit(Xtr,Ytr)  #fitting my data into random_search_lr results

print('best score {}'.format(random_search_lr.best_score_)) #printing best score
print('best hyperparameter {}'.format(random_search_lr.best_params_)) #printing best hyperparameters

Fitting 1 folds for each of 10 candidates, totalling 10 fits


C:\Users\Debi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:922: UserWarning: One or more of the test scores are non-finite: [0.84789151        nan        nan        nan        nan 0.85562872
        nan        nan        nan 0.68890488]
  warnings.warn(


best score 0.8556287210249962
best hyperparameter {'tfidf__ngram_range': (1, 4), 'tfidf__min_df': 16, 'tfidf__max_df': 0.6, 'lr__solver': 'saga', 'lr__penalty': 'l1', 'lr__C': 58780.16072274924}


C:\Users\Debi\anaconda3\lib\site-packages\sklearn\linear_model\_sag.py:328: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn("The max_iter was reached which means "


In [25]:
submission = pd.DataFrame() # creating dataframe for the submission

submission['id'] = test_data_clean.index #using the column index of x test values to fill submission['id'] column

submission['label'] = random_search_lr.predict_proba(test_data_clean)[:,1] #predicting the probabilities and filling submission['label'] column with their values

submission.to_csv('predwalkthrough.csv', index=False)#generating the submission file


# Trial two using logistic regression classifier with random search and validation set but with different hyperparameters ranges

# Expectaion

I expected to get a higher score after changing the ranges of my hyper parammetrs

# Observation 

After running the code,the best hyperparameters that was defined were : 

'tfidf__ngram_range': (1, 4)

'tfidf__min_df': 16

'tfidf__max_df': 0.6

'lr__solver': 'sag'

'lr__penalty': 'l1'

'lr__C': 58780.16072274924

kaggle score 0.83286

best score 0.8556287210249962
    

# Plan

I'm going to use the xgboost model

In [26]:
import xgboost as xgb


In [27]:
pipe = Pipeline([("tfidf", TfidfVectorizer("char")),  ('xgb_model',xgb.XGBClassifier())])

params = {
    
    "tfidf__ngram_range": [(1, 4),(2,5)], # #n-gram range:tuple (min_n, max_n), The lower and upper boundary of the range of n-values for different n-grams to be extracted. All values of n such that min_n <= n <= max_n will be used.
    "tfidf__max_df": np.arange(0.6, 0.8),#When building the vocabulary ignore terms that have a document frequency strictly higher than the given threshold (corpus-specific stop words).
    "tfidf__min_df": np.arange(5,30),#When building the vocabulary ignore terms that have a document frequency strictly lower than the given threshold. This value is also called cut-off in the literature. 
    'xgb_model__min_child_weight': [1, 5, 10], #The number of samples required to form a leaf node (the end of a branch)
    'xgb_model__gamma': [0.5, 1, 1.5, 2, 5],#The gamma is an unbounded parameter from 0 to infinity that is used to control the model’s tendency to overfit.
    'xgb_model__subsample': [0.6, 0.8, 1.0], # determines how much of the initial dataset is fair game for random sampling during each iteration of the boosting process.
    'xgb_model__colsample_bytree': [0.6, 0.8, 1.0],#defines percentage of features will be used for building each tree.
    'xgb_model__max_depth': [3, 4, 5]# how deep each estimator is permitted to build a tree.
}

C:\Users\Debi\anaconda3\lib\site-packages\sklearn\utils\validation.py:70: FutureWarning: Pass input=char as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error
  warnings.warn(f"Pass {args_msg} as keyword args. From version "


In [28]:
random_search_xg = RandomizedSearchCV(
    pipe, params, cv=pds, verbose=1, n_jobs=2, 
    # number of random trials
    n_iter=10,
    scoring='roc_auc')#The degree of separability/distinction or intermingling/crossover between the forecasts of the two classes is shown by the ROC-AUC.

random_search_xg.fit(Xtr,Ytr) #fitting my x and y into the random search results

print('best score {}'.format(random_search_xg.best_score_)) #printing  best scores
print('best hyperparameter {}'.format(random_search_xg.best_params_)) #printing best hyperparameters

Fitting 1 folds for each of 10 candidates, totalling 10 fits


C:\Users\Debi\anaconda3\lib\site-packages\xgboost\sklearn.py:1224: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[19:57:41] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
best score 0.8206981714790174
best hyperparameter {'xgb_model__subsample': 0.8, 'xgb_model__min_child_weight': 1, 'xgb_model__max_depth': 4, 'xgb_model__gamma': 5, 'xgb_model__colsample_bytree': 1.0, 'tfidf__ngram_range': (1, 4), 'tfidf__min_df': 16, 'tfidf__max_df': 0.6}


In [29]:
submission = pd.DataFrame() # creating dataframe for the submission

submission['id'] = test_data_clean.index #using the column index of x test values to fill submission['id'] column

submission['label'] = random_search_xg.predict_proba(test_data_clean)[:,1] #predicting the probabilities and filling submission['label'] column with their values

submission.to_csv('predwalkthrough.csv', index=False)#generating the submission file

# Trial three using xgboost model with random search and validation set  

# Expectaion

I expected to get a higher score after changing the model

# Observation 

After running the code,the best hyperparameters that was defined were : 

'xgb_model__subsample': 0.8

'xgb_model__min_child_weight': 1

'xgb_model__max_depth': 4

'xgb_model__gamma': 5

'xgb_model__colsample_bytree': 1.0

'tfidf__ngram_range': (1, 4)

'tfidf__min_df': 16

'tfidf__max_df': 0.6

kaggle score Score: 0.77052

best score 0.8206981714790174

# Plan

I will change the hyperparameters' ranges to the xgboost model and I will use the word level vectorizer

In [30]:
pipe = Pipeline([("tfidf", TfidfVectorizer("word")),  ('xgb_model',xgb.XGBClassifier())])

params = {
    
    "tfidf__ngram_range": [(3, 6),(4,8)],#n-gram range:tuple (min_n, max_n), The lower and upper boundary of the range of n-values for different n-grams to be extracted. All values of n such that min_n <= n <= max_n will be used.
    "tfidf__max_df": np.arange(0.3, 0.9),#When building the vocabulary ignore terms that have a document frequency strictly higher than the given threshold (corpus-specific stop words).
    "tfidf__min_df": np.arange(8,24),#When building the vocabulary ignore terms that have a document frequency strictly lower than the given threshold. This value is also called cut-off in the literature.
    'xgb_model__min_child_weight': [3, 6, 11], #The number of samples required to form a leaf node (the end of a branch)
    'xgb_model__gamma': [0.25, 0.75, 2, 2.5, 5],#The gamma is an unbounded parameter from 0 to infinity that is used to control the model’s tendency to overfit.
    'xgb_model__subsample': [0.2, 0.5, 0.6], # determines how much of the initial dataset is fair game for random sampling during each iteration of the boosting process.
    'xgb_model__colsample_bytree': [0.6, 0.8, 1.0],#defines percentage of features will be used for building each tree.
    'xgb_model__max_depth': [3, 6, 9]# how deep each estimator is permitted to build a tree.
}

C:\Users\Debi\anaconda3\lib\site-packages\sklearn\utils\validation.py:70: FutureWarning: Pass input=word as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error
  warnings.warn(f"Pass {args_msg} as keyword args. From version "


In [31]:
random_search_xg = RandomizedSearchCV(
    pipe, params, cv=pds, verbose=1, n_jobs=2, 
    # number of random trials
    n_iter=10,
    scoring='roc_auc')

random_search_xg.fit(Xtr,Ytr) #fitting my x and y into the xgboost random search rresults

print('best score {}'.format(random_search_xg.best_score_)) #printing best scores
print('best hyperparameter {}'.format(random_search_xg.best_params_)) #printing best hyperparameters

Fitting 1 folds for each of 10 candidates, totalling 10 fits
[19:58:07] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
best score 0.5076353269477214
best hyperparameter {'xgb_model__subsample': 0.6, 'xgb_model__min_child_weight': 6, 'xgb_model__max_depth': 6, 'xgb_model__gamma': 2, 'xgb_model__colsample_bytree': 0.8, 'tfidf__ngram_range': (3, 6), 'tfidf__min_df': 18, 'tfidf__max_df': 0.3}


In [32]:
submission = pd.DataFrame() # creating dataframe for the submission

submission['id'] = test_data_clean.index #using the column index of x test values to fill submission['id'] column

submission['label'] = random_search_xg.predict_proba(test_data_clean)[:,1] #predicting the probabilities and filling submission['label'] column with their values

submission.to_csv('predwalkthrough.csv', index=False)#generating the submission file

# Trial four using xgboost model with random search and validation set but with different hyperparameters ranges

# Expectaion

I expected to get a higher score after changing the hyper parameters' ranges

# Observation 

After running the code,the best hyperparameters that was defined were : 

'xgb_model__subsample': 0.6

'xgb_model__min_child_weight': 6

'xgb_model__max_depth': 6

'xgb_model__gamma': 2

'xgb_model__colsample_bytree': 0.8

'tfidf__ngram_range': (3, 6)

'tfidf__min_df': 18

'tfidf__max_df': 0.3

kaggle score Score: 0.50315

best score 0.5076353269477214

# Plan

I will change my model to the randomforest () and I will use char level vectorizer

In [33]:
from sklearn.ensemble import RandomForestClassifier

pipe = Pipeline([("tfidf", TfidfVectorizer("word")),  ('my_classifier',RandomForestClassifier())])

params = {
    
    "tfidf__ngram_range": [(3, 6),(4,8)],#n-gram range:tuple (min_n, max_n), The lower and upper boundary of the range of n-values for different n-grams to be extracted. All values of n such that min_n <= n <= max_n will be used.
    "tfidf__max_df": np.arange(0.3, 0.9),#When building the vocabulary ignore terms that have a document frequency strictly higher than the given threshold (corpus-specific stop words).
    "tfidf__min_df": np.arange(8,24),#When building the vocabulary ignore terms that have a document frequency strictly lower than the given threshold. This value is also called cut-off in the literature.
    'my_classifier__n_estimators': [50, 100, 150],  # ranges of n_estimators which are the number of trees to be used in the forest.
    # I set my  n_estimators ranges to [50, 100, 150]
    'my_classifier__max_depth':[30, 60, 90]  
    # I set my  max depth ranges to [30, 60, 90] which are The number of splits that each decision tree is allowed to make.
}

In [34]:
random_search_rnf = RandomizedSearchCV(
    pipe, params, cv=pds, verbose=1, n_jobs=2, 
    # number of random trials
    n_iter=10,
    scoring='roc_auc')

random_search_rnf.fit(Xtr,Ytr)

print('best score {}'.format(random_search_rnf.best_score_))
print('best hyperparameter {}'.format(random_search_rnf.best_params_))

Fitting 1 folds for each of 10 candidates, totalling 10 fits
best score 0.5185222470697569
best hyperparameter {'tfidf__ngram_range': (3, 6), 'tfidf__min_df': 11, 'tfidf__max_df': 0.3, 'my_classifier__n_estimators': 100, 'my_classifier__max_depth': 60}


In [35]:
submission = pd.DataFrame() # creating dataframe for the submission

submission['id'] = test_data_clean.index #using the column index of x test values to fill submission['id'] column

submission['label'] = random_search_rnf.predict_proba(test_data_clean)[:,1] #predicting the probabilities and filling submission['label'] column with their values

submission.to_csv('predwalkthrough.csv', index=False)#generating the submission file

# Trial five using randomforest() model with random search and validation set 

# Expectaion

I expected to get a higher score after changing the model

# Observation 

After running the code,the best hyperparameters that was defined were : 

'tfidf__ngram_range': (3, 6)

'tfidf__min_df': 11

'tfidf__max_df': 0.3

'my_classifier__n_estimators': 100

'my_classifier__max_depth': 60

score on kaggle :  0.50691

best score 0.5185222470697569

# Refrences

https://pdf.sciencedirectassets.com/277811/1-s2.0-S1877042813X0028X/1-s2.0-S1877042813041918/main.pdf?X-Amz-Security-Token=IQoJb3JpZ2luX2VjEOb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJHMEUCIQCjejOQyKdO7U33cK6CHIkH%2F5RcEs9esfrECmpPNqnZxwIgDXTl6duLcDQMHQ9xBtE%2BGp4O9%2B33gQRMH2yhqFfCRokq%2BgMIXhAEGgwwNTkwMDM1NDY4NjUiDPcCm7JZf8j%2F2SKTLirXA41KPsw6b3nl8BuptqiheyReIhuIAACodyueVBEn3xA977Noz2ZIbVQgnOlzvQ8jhgvOgqk%2FN6XjbVN7ErmCqffpY1p%2FO31NoiKT7zKoNw9k11blK1hPNt%2B%2FpnH926DnCfVP%2FxqvFvqRM%2FjWJwL2ko%2B6iiIZrrbmbToCVGcbNxLC1lNjeZEAhDn%2BWg1qfmnOd6xROkRa1LtTN13XJKBE3DYpQkHsIkO%2B1bjqzqKmgGZ6F2FWFdTgFB0bfLR2QoOGKoSLi1psJ%2FVKO1ODMZNIkjdnJgjEfzyw09Zl80qA%2F0FufLouJFTkwnLJGDRfH5VtRVyHVTs8nZ7q7j2Pg%2BuCygnfst0blSsHVgiddtQwW302dYDHlSw15%2FixfPJkNgGI9Cr6tu4yi3mnsg9MitXzAi8Hr9JWkMoqhFePKuSx%2FVlCW0OVjmecjuEgH7mU4Lv5LflUFSQbvbuJ07D%2BsHpfMk%2BxvBAu2DUaxxkrVofZSHX%2BDmB0Jy7C2ES5If9ov6HilVHIYDXP6YW4QacHuOV90KEhB%2FRnajd1I8OA%2BvZbtPay318M%2FntL8NLOPFCLXnZtwKH1zC54PQS1kSL8RLJSeIQ%2BEUw04sk1FhzOjzDjPnWPMGCUkhi%2F%2FzC7%2FbyRBjqlASY0PYfL9%2FFb3BPo%2F3F9AG9NvNQY%2BFC2vSvoX7cwe0X1mNW3S2ntFwNRahsiDYk%2BkM4%2FwcwPf%2B9ATO0%2BA1ityvdAdFmK5YEERp09H7b2bXJWfBc8uY2udNQEvsHOscNwnjZtL9jX5QQeGVtxSJzD4Ptu61FVgvP0S315XMrTu8GvyaDQ45PQAUIbR849GvUAVm3Hw7pUx9Yn4jHS%2FFfSV5MP0ar0Ag%3D%3D&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20220314T133850Z&X-Amz-SignedHeaders=host&X-Amz-Expires=300&X-Amz-Credential=ASIAQ3PHCVTYUPZR4G5W%2F20220314%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=da106f5b7265022db1ea709957cd21b63b435910ca1fd39fffaa848dacca37fa&hash=f595daebc10953250ebbd13f1925d859be6341035ceacba5fbea2152b96975b8&host=68042c943591013ac2b2430a89b270f6af2c76d8dfd086a07176afe7c76c2c61&pii=S1877042813041918&tid=spdf-3cf4728f-3ded-491c-ae14-58a482dd28c7&sid=cd3ce52c85b4774e1a0a6db1bb4f7b548184gxrqb&type=client&ua=4c00030b595b50535f&rr=6ebd69761e905fc5

https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html

https://charmestrength.com/what-is-penalty-in-logistic-regression/

https://kevinvecmanis.io/machine%20learning/hyperparameter%20tuning/dataviz/python/2019/05/11/XGBoost-Tuning-Visual-Guide.html

https://datascience.stackexchange.com/questions/25581/what-is-the-difference-between-countvectorizer-token-counts-and-tfidftransformer#:~:text=The%20only%20difference%20is%20that%20the%20TfidfVectorizer%20%28%29,%28%29%20assigns%20a%20score%20while%20CountVectorizer%20%28%29%20counts.

https://nlp.stanford.edu/IR-book/html/htmledition/tokenization-1.html#:~:text=This%20guarantees%20that%20a%20sequence,the%20document%20to%20be%20known.

https://www.analyticsvidhya.com/blog/2021/09/what-are-n-grams-and-how-to-implement-them-in-python/

https://help.monkeylearn.com/en/articles/2174105-n-gram-range